### PR TITLE
regroup taint management in one place to easy code maintenance and understanding

### PR DESCRIPTION
**Pull request by dbenque**
Taint management is done in only one place in the handleCandidate function.

### CODE REVIEW

File: internal/drain_runner/runner.go [hunk 1]
```diff
@@ -116,51 +122,82 @@ func (runner *drainRunner) handleGroup(ctx context.Context, info *groups.RunnerI
...
+	if err = runner.drainCandidate(ctx, info, candidate); err != nil {
+		loggerForNode.Error(err, "failed to drain node")
 		runner.eventRecorder.NodeEventf(ctx, candidate, core.EventTypeWarning, kubernetes.EventReasonDrainFailed, "Drain failed: %v", err)
-		return runner.removeFailedCandidate(ctx, candidate, err.Error())
+		updatedNode, errRetryWall := runner.updateRetryWallOnCandidate(ctx, candidate, err.Error())
+		if errRetryWall != nil {
+			loggerForNode.Error(errRetryWall, "Failed to remove taint following drain failure")
+			return err
+		}
+		if _, errTaint := k8sclient.RemoveNLATaint(ctx, runner.client, updatedNode); err != nil {
```
Comment (dhenkel92) [inline internal/drain_runner/runner.go:169]:
```suggestion
		if _, errTaint := k8sclient.RemoveNLATaint(ctx, runner.client, updatedNode); errTaint != nil {
```

File: internal/drain_runner/runner.go [hunk 2]
```diff
@@ -116,51 +122,82 @@ func (runner *drainRunner) handleGroup(ctx context.Context, info *groups.RunnerI
...
+	runner.eventRecorder.NodeEventf(ctx, candidate, core.EventTypeNormal, kubernetes.EventReasonDrainStarting, "Draining node")
+	// TODO add metric to show how many drains are successful / failed
+	if err = runner.drainCandidate(ctx, info, candidate); err != nil {
+		loggerForNode.Error(err, "failed to drain node")
 		runner.eventRecorder.NodeEventf(ctx, candidate, core.EventTypeWarning, kubernetes.EventReasonDrainFailed, "Drain failed: %v", err)
-		return runner.removeFailedCandidate(ctx, candidate, err.Error())
+		updatedNode, errRetryWall := runner.updateRetryWallOnCandidate(ctx, candidate, err.Error())
+		if errRetryWall != nil {
+			loggerForNode.Error(errRetryWall, "Failed to remove taint following drain failure")
+			return err
```
Comment (dhenkel92) [inline internal/drain_runner/runner.go:167]:
```suggestion
			return errRetryWall
```

File: internal/drain_runner/runner.go [hunk 3]
```diff
@@ -116,51 +122,82 @@ func (runner *drainRunner) handleGroup(ctx context.Context, info *groups.RunnerI
...
+		}
+		if _, errTaint := k8sclient.RemoveNLATaint(ctx, runner.client, updatedNode); err != nil {
+			loggerForNode.Error(errTaint, "Failed to remove taint following drain failure")
+			return errTaint
+		}
+		return err
 	}
+	candidate, err = k8sclient.AddNLATaint(ctx, runner.client, candidate, runner.clock.Now(), k8sclient.TaintDrained)
```
Comment (dhenkel92) [inline internal/drain_runner/runner.go:175]: error is not used

File: internal/drain_runner/runner.go [hunk 4]
```diff
@@ -86,10 +87,15 @@ func (runner *drainRunner) handleLeftOverDraining(ctx context.Context, info *gro
 	}
 	// TODO add metric to track amount of nodes blocked in draining
 	for _, n := range draining {
-		if errRmTaint := runner.removeFailedCandidate(ctx, n, "Node stuck in draining (controller restart?)"); errRmTaint != nil {
+		updatedNode, errRmTaint := runner.updateRetryWallOnCandidate(ctx, n, "Node stuck in draining (controller restart?)")
+		if errRmTaint != nil {
 			// we just log the error, it will come back at next iteration
 			runner.logger.Error(err, "Failed to remove 'draining' taint", "node", n.Name)
```
Comment (dhenkel92) [inline internal/drain_runner/runner.go:93]: Please update message

File: internal/drain_runner/runner.go (hunk 1 above)
Comment (dbenque) [reply internal/drain_runner/runner.go:169]: 👍

File: internal/drain_runner/runner.go (hunk 2 above)
Comment (dbenque) [reply internal/drain_runner/runner.go:167]: 👍

File: internal/drain_runner/runner.go (hunk 4 above)
Comment (dbenque) [reply internal/drain_runner/runner.go:93]: done

File: internal/drain_runner/runner.go (hunk 1 above)
Comment (dhenkel92) [reply internal/drain_runner/runner.go:186]: You forgot this one 😅

File: internal/drain_runner/runner.go [hunk 5]
```diff
@@ -181,15 +181,19 @@ func (runner *drainRunner) handleCandidate(ctx context.Context, info *groups.Run
...
+			return errRetryWall
 		}
 		if _, errTaint := k8sclient.RemoveNLATaint(ctx, runner.client, updatedNode); err != nil {
 			loggerForNode.Error(errTaint, "Failed to remove taint following drain failure")
 			return errTaint
 		}
 		return err
 	}
-	candidate, err = k8sclient.AddNLATaint(ctx, runner.client, candidate, runner.clock.Now(), k8sclient.TaintDrained)
+	if candidate, err = k8sclient.AddNLATaint(ctx, runner.client, candidate, runner.clock.Now(), k8sclient.TaintDrained); err != nil {
```
Comment (dhenkel92) [inline internal/drain_runner/runner.go:192]: Is this overwriting the variables even out of the if scope?

Comment (dhenkel92) [reply internal/drain_runner/runner.go:192]: I did a small test where it worked 🙂

Comment (dbenque) [reply internal/drain_runner/runner.go:192]: yes it is since the definition is out of this scope.

**Review by dhenkel92 — APPROVED** [review]

**Pull request merged**